### PR TITLE
IA-1648 Generalize xls form generation from polio campaign to iaso globally

### PR DESCRIPTION
**Pull request by SimonDus**
Explain what problem this PR is resolving :
Export a xlsform by using a group as OU source. A template is required. Specific data can be extracted from campaigns by
starting the variable name by 'insert_' in the xls file in "calculate" row. The data will be saved in the
calculation column. By adding campaignid to the params you can extract data from a campaign.

## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] My migrations file are included
- [x] Are there enough tests

## How to test
The url to test it with extracting data from a campaign looks like this 
/api/generate_xlsform/xlsform_generator/?groupid=169&form_name=test_form&campaignid=022ca298-b24c-479b-ad81-34a4a6e9782b&ouhierarchy=COUNTRY,REGION,district&outypelist=country,region,district

- groupid  : It's the group of org units that we want to use to generate the xls form. The user can only access the groups linked to his account. 
- form_name : it's the template form name.
- campaignid : the campaign id must be specified in order to extract data from a campaign
- ouhierarchy : provide a list where i[0] is the top parent i[1] is the children i[2] and so on...
- outypelist: list of the type of orgunit we want in the xlsform. 

A xls form template is required.  The xls must contains the basics required field of an xlsform. The choice sheet will be generated. 

The column name for the survey sheet are :
type | name | label | required | appearance | constraint | constraint_message | relevant | hint | choice_filter | jr:count | repeat_count | calculation

Check below for a video on how to test it. 

https://user-images.githubusercontent.com/45785235/204535954-27bed13e-9455-41f0-9c91-74ed2c1b6001.mov

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ geopy==1.21.0
 Shapely==1.7.0
 unidecode==1.2.0
 drf-yasg==1.20.0
+openpyxl==3.0.10
```
Comment (olethanh) [inline requirements.txt]: it is there twice?

File: plugins/polio/models.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from typing import Union
+import pathlib
```
Comment (olethanh) [inline plugins/polio/models.py]: probably a left over, shouldn't be modified

File: plugins/polio/api.py [hunk 3]
```diff
@@ -21,6 +20,7 @@
 from django.utils.timezone import now, make_aware
 from django_filters.rest_framework import DjangoFilterBackend  # type: ignore
 from gspread.utils import extract_id_from_url  # type: ignore
+from openpyxl.writer.excel import save_virtual_workbook
```
Comment (olethanh) [inline plugins/polio/api.py]: idem

File: iaso/api/xls_form_generator.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+from rest_framework import serializers, filters
+from rest_framework.viewsets import ModelViewSet
+import string
+from datetime import datetime
+from tempfile import NamedTemporaryFile
+import openpyxl
+from iaso.api.common import DeletionFilterBackend, TimestampField
+from iaso.models import Group
+from iaso.models.xls_form_template import XlsFormTemplate
+from plugins.polio.models import Campaign
```
Comment (olethanh) [inline iaso/api/xls_form_generator.py]: if you import model from Polio in the main iaso code this is making iaso crash when you run it without the polio plugins activated.

Probably the Campaign related logic could be moved to the polio plugins itself?
If you still need to do an import somehow, you will need to wrap it in a try except

File: iaso/api/xls_form_generator.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+class XlsFormTemplateSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = XlsFormTemplate
+        fields = "__all__"
+
+    created_at = TimestampField(read_only=True)
+    updated_at = TimestampField(read_only=True)
+
+
+def get_data_from_campaigns(campaign_id, row, q_sheet, calculation_index):
```
Comment (niconoe) [inline iaso/api/xls_form_generator.py]: Could we add type hints there?

File: iaso/api/xls_form_generator.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+        filename = f"FORM_{form_name}_{datetime.now().date()}.xlsx"
+
+        with NamedTemporaryFile() as tmp:
+            wb.save(tmp.name)
+            tmp.seek(0)
+            stream = tmp.read()
+
+            response = HttpResponse(
+                stream, content_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
```
Comment (niconoe) [inline iaso/api/xls_form_generator.py]: Use `iaso.api.common.CONTENT_TYPE_XLSX` instead

File: iaso/api/xls_form_generator.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+    def get_serializer_class(self):
+        return XlsFormTemplateSerializer
+
+    def get_queryset(self):
+        queryset = XlsFormTemplate.objects.filter(account=self.request.user.iaso_profile.account)
+        return queryset
+
+    @action(methods=["GET"], detail=False)
+    def xlsform_generator(self, request) -> HttpResponse:
```
Comment (niconoe) [inline iaso/api/xls_form_generator.py:219]: This method is really long and focus on many things (building the request, creating a XLS file, moving data to it, ...). Would it be possible to refactor so it's shorter and more readable?

File: iaso/models/org_unit.py [hunk 8]
```diff
@@ -341,6 +345,9 @@ def calculate_paths(self, force_recalculate: bool = False) -> typing.List["OrgUn
     def __str__(self):
         return "%s %s %d" % (self.org_unit_type, self.name, self.id if self.id else -1)
 
+    def create_ou_tree(self):
```
Comment (niconoe) [inline iaso/models/org_unit.py:348]: Can we add a docstring explaining what it does, a type annotation for a return value and a unit test to make sure it works (and continue to work in the future)?

File: iaso/tests/api/test_xls_form_generator.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+from iaso.models.xls_form_template import XlsFormTemplate
+
+from django.utils.timezone import now
+
+from iaso import models as m
+from iaso.test import APITestCase
+
+
+class XlsFormGeneratorAPITestCase(APITestCase):
```
Comment (niconoe) [inline iaso/tests/api/test_xls_form_generator.py:14]: Can we also test the actual content of the generated file?

**Review by olethanh — COMMENTED** [review]
haven't test it yet, some preleminary remarks from having read the code diagonally

File: plugins/polio/api.py (hunk 3 above)
Comment (SimonDus) [reply plugins/polio/api.py]: this one is actually required.

**Comment by SimonDus** [comment]
@niconoe @olethanh I did the requested changes

**Comment by mathvdh** [comment]
Je comprends pas trop pourquoi ce truc est tellement énorme ?

**Comment by mathvdh** [comment]
My bad, github a dû bugger ou quelque chose parce qu'il me montrait genre 1649 fichiers dans la PR ..
:exploding_head:  
Maintenant ça ne le fais plus, je vois 22 fichiers changés

File: iaso/api/xls_form_generator.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    return stream, filename
+
+
+def get_data_from_campaigns(campaign_id: str, row: int, q_sheet: Worksheet, calculation_index: int):
+    """This function get the data from a campaign and insert it in the calculation row
+    of the xlsform. It's tagged as 'insert_field' in the xlsform.
+    """
+    authorized_fields = [
+        "id",
+        "epid",
```
Comment (madewulf) [inline iaso/api/xls_form_generator.py:135]: Old PR I know, but I don't get why we still list the authorized fields if we want to pull the data from the api before sending it back to the generation API

**Review by mathvdh — APPROVED** [review]
For me its fine. There is a hardcode list of fields but we couldn't find a easy solution for this.

**Pull request closed without merging**